tests/kostentraeger/transformer.spec.ts

```diff
@@ -0,0 +1,198 @@
+                ik: "999999999",
+                abbreviatedName: "short name",
+                name: "very long name",
+                validityFrom: new Date("2020-20-20"),
+                validityTo: new Date("2088-10-10"),
+                bankAccountDetails: {
+                    bankName: "Sparbank",
+                    accountOwner: "short name",
+                    iban: "ibanibaniban",
+                    bic: "bicbicbic"
+                },
+                addresses: [
+                    { postcode: 12345, place: "Humburg", streetAndHousenumber: "Straßenallee 33" },
+                    { postcode: 12345, place: "Humburg", poBox: "123" },
+                    { postcode: 12345, place: "Humburg" },
+                ],
+                contacts: [
+                    {
+                        phone: "123456789/88",
+                        fax: "123456789/77",
+                        name: "Max & Moritz",
+                        fieldOfWork: "Schabernack"
+                    }
+                ],
+                transmissionMethods: {
+                    paperReceipt: false,
+                    machineReadablePaperReceipt: true,
+                    email: "ok@go.de",
+                    ftam: "ftam.blub-it.de:5000",
+                    zeichensatzSchluessel: "I8"
+                },
+                links: [
+                    {
+                        ikVerknuepfungsartSchluessel: "01",
+                        verknuepfungspartnerIK: "555444333",
+                        leistungserbringergruppeSchluessel: "6",
+                        datenlieferungsartSchluessel: "07",
+                        standortLeistungserbringerBundeslandSchluessel: "02",
+                        pflegeLeistungsartSchluessel: "00"
+                    },
+                    {
+                        ikVerknuepfungsartSchluessel: "03",
+                        verknuepfungspartnerIK: "334455667",
+                        leistungserbringergruppeSchluessel: "5",
+                        datenlieferungsartSchluessel: "07",
+                        standortLeistungserbringerKVBezirkSchluessel: "37",
+                        abrechnungscodeSchluessel: "25",
+                        tarifkennzeichen: "12345"
+                    },
+                    {
+                        ikVerknuepfungsartSchluessel: "09",
+                        verknuepfungspartnerIK: "112233445",
+                        leistungserbringergruppeSchluessel: "6",
+                        datenlieferungsartSchluessel: "29",
+                        pflegeLeistungsartSchluessel: "00"
+                    }
+                ]
+            }],
+        }
+
+        /* need to compare the stringified and then parsed result because Javascript Date objects 
+           are compared using identity, not equality :-( */
+        const actual = JSON.parse(JSON.stringify(transform(interchange)))
+        const expected = JSON.parse(JSON.stringify(expectedInstitutionList))
```

Comment:
Probably it is of no importance in this case. So just in case, I learned the hard way that when serializing to JSON, Date objects are converted to strings, which remain strings when parsed back to JSON, unless you use a replace / reviver function as parameter.

Reply:
Good to know. However in this case this is deliberate.

Though I'll latest stumble over it if I stringify the kostenträger-information somewhere and then parse it back into memory. How'd a reviver function look like? Should then the data model perhaps not include the parsed dates but just the ISO timestamps as strings?

Reply:
JSON with a Date object automatically looks like this when stringified: `const json = "{\"createdAt\":\"2021-04-27T08:20:30.507Z\"}"`

The JSON parser calls the reviver function for every key value pair. A simple reviver function would check if the value is of type string and if the string matches the date format of a regular expression to convert the string back to a Date object:

```
const dateStringRegexp = /(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})/

const result = JSON.parse(json, (key, value) => {
    if (typeof value == "string" && (dateStringRegexp.exec(value))) {
        return new Date(value);
    } else {
        return value;
    }
})
```

Concerning the second question: I think we can keep the model as it is, although a JavaScript object containing Date objects is actually not valid JSON. We just need to keep in mind to use a reviver function in case we need to serialize the object as string to persist it. I can't think of a reason that would speak against it right now.

Reply:
ok